### ISSUE TITLE
Arcade does not support /p:OptionalToolSource options

### ISSUE THREAD

**Issue by adiaaida** [dotnet/arcade#1176]
In the corefx build, we pulled down the OptionalToolSource packages using sync.cmd /p:OptionalToolSource, /p:OptionalToolSourceUser and /p:OptionalToolSourcePassword. With the port to Arcade, these options no longer pull down the optional tools, which are required for ProfileGuidedOptimization in corefx (specifically, applying IBC data to the various assemblies).

**Comment by maririos** [comment]
FYI @weshaggard

**Comment by dagood** [comment]
Arcade support is tracked in https://github.com/dotnet/arcade/issues/100. It sounds like this issue does more generally track getting the flow working in CoreFX though... I would have thought that the BuildTools code that AFAIK is still in CoreFX would be used to keep support for this.

**Comment by weshaggard** [comment]
It looks like the BuildTools optional tooling stuff hooked onto the Sync target (https://github.com/dotnet/buildtools/blob/0fa455c550abb4f49f4830a7f2fa78609f373fd0/src/Microsoft.DotNet.Build.Tasks/PackageFiles/OptionalTooling.targets#L12) I deleted that target from corefx recently (https://github.com/dotnet/corefx/commit/66392f577c7852092f668876822b6385bcafbd44#diff-3ad318a39efd856540e8cf813f20ec09) and renamed it to Restore to align with arcade targets. I'll submit a change in corefx to hook up that target again.

**Comment by weshaggard** [comment]
https://github.com/dotnet/corefx/pull/33076 should hook up the optional tooling targets again.

**Comment by tmat** [comment]
This is how we handle IBCMerge in Roslyn and msbuild:

https://github.com/dotnet/roslyn/blob/master/.vsts-ci.yml#L29-L32
https://github.com/dotnet/roslyn/blob/master/build/ToolsetPackages/InternalToolset.csproj

**Comment by weshaggard** [comment]
@tmat how to you handle the credentials for the feed? Is that hidden away in the feed guid you are using?

**Comment by tmat** [comment]
Yes, the DevOps build task runs with credentials that can access that guid feed.

**Comment by tmat** [comment]
What is the local dev scenario for merging optimization data?

**Comment by tmat** [comment]
Do you mean just replicating potential failure locally? If so then yes, it's not hard to get the package from the VS feed.

**Comment by adiaaida** [comment]
There really isn't much of a local dev scenario for merging optimization data, other than for testing new functionality before checking in (like how @DrewScoggins is working on getting IBC data processed for Linux jobs). The only place that optimization data is consumed for CoreFx is in the official build.

**Comment by markwilkie** [comment]
Perhaps yet another opportunity for a AzDO tasks?

cc/ @chcosta

**Comment by tmat** [comment]
I don't think we need a special task for this. The NuGet one already handles it just fine.

**Comment by chcosta** [comment]
A special task wouldn't address the "local dev scenario" @dagood brings up.  I'm also not sure that a task addresses the fact that most of these packages need to get plugged into the build.  It does seem that Arcade could perhaps provide an extensibility point here though which would provide these packages if certain authentication requirements were met.  That would allow an official build or a local dev to repro using the same mechanisms.  @tmat, thoughts?

**Comment by tmat** [comment]
I'm not sure what problem are we trying to solve here. The current solution works well enough. We can probably improve on it in future, once https://github.com/NuGet/Home/issues/6045 is fixed.

**Comment by dagood** [comment]
For local dev scenario, I think all I'm asking for is a pointer to a doc.

YAML templating or inclusion in some base template extensibility sounds fine. (Although it doesn't seem like anyone's really asking for that.) A custom-built task doesn't sound very useful though--I don't see what would/could be improved over the four-liner NuGet Restore task tmat's already pointed out in both issues tracking this feature.

**Comment by chcosta** [comment]
I don't think this would be a YAML template thing, I hadn't seen that proposed or even considered it.

**Comment by dagood** [comment]
Ah, sorry, that's what I interpreted `It does seem that Arcade could perhaps provide an extensibility point here though which would provide these packages if certain authentication requirements were met.` to mean--I was thinking too hard about it in context of build steps.

For the Arcade SDK, that would be similar to the BuildTools approach. What I'm thinking @tmat has said is that we shouldn't need that additional tooling for local devs to find this easy to use. (Either way, I'm just hoping for a doc with steps that are straightforward to follow, extra tooling or no.)

**Comment by tmat** [comment]
@dagood This is the command to use to restore the internal toolset project:

```
D:\Roslyn\build\ToolsetPackages>c:\programs\NuGet\nuget restore InternalToolset.csproj -Source https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/index.json
```

```
MSBuild auto-detection: using msbuild version '15.9.14.22190' from 'C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\MSBuild\15.0\bin'.
Restoring packages for D:\Roslyn\build\ToolsetPackages\InternalToolset.csproj...
CredentialProvider.VSS: Getting new credentials for source:https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/index.json, scope:vso.packaging_write vso.drop_write
  GET https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/flat2/microsoft.dotnet.ibcmerge/index.json
  OK https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/flat2/microsoft.dotnet.ibcmerge/index.json 208ms
  GET https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/flat2/microsoft.dotnet.ibcmerge/4.7.2-alpha-00001/microsoft.dotnet.ibcmerge.4.7.2-alpha-00001.nupkg
  OK https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/flat2/microsoft.dotnet.ibcmerge/4.7.2-alpha-00001/microsoft.dotnet.ibcmerge.4.7.2-alpha-00001.nupkg 674ms
Installing Microsoft.DotNet.IBCMerge 4.7.2-alpha-00001.
Committing restore...
Generating MSBuild file D:\Roslyn\build\ToolsetPackages\obj\InternalToolset.csproj.nuget.g.props.
Writing lock file to disk. Path: D:\Roslyn\build\ToolsetPackages\obj\project.assets.json
Restore completed in 7.04 sec for D:\Roslyn\build\ToolsetPackages\InternalToolset.csproj.

NuGet Config files used:
    D:\Roslyn\NuGet.Config
    C:\Users\tomat\AppData\Roaming\NuGet\NuGet.Config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

Feeds used:
    https://devdiv.pkgs.visualstudio.com/_packaging/8f470c7e-ac49-4afe-a6ee-cf784e438b93/nuget/v3/index.json

Installed:
    1 package(s) to D:\Roslyn\build\ToolsetPackages\InternalToolset.csproj
```

**Comment by maririos** [comment]
@tmat is the `internal toolset project` part of Arcade? or something that will be ported to Arcade that everyone needs to know? Or is it something only Roslyn does.

@dagood it looks like the conversation has drifted from the original ask of the issue. Do you mind opening an issue with your request for a doc?

**Comment by weshaggard** [comment]
We should move this conversation to https://github.com/dotnet/arcade/issues/100.
